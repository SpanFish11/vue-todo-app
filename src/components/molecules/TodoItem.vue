<script setup>
import AppCheckboxInput from '@/components/atoms/AppCheckboxInput.vue'
import AppButton from '@/components/atoms/AppButton.vue'

const props = defineProps({
  name: { type: String, required: true },
  completed: { type: Boolean, default: false },
})
</script>

<template>
  <div class="todo-item">
    <AppCheckboxInput :checked="props.completed" />
    <div>
      <span
        :class="{
          'todo-item-completed': props.completed,
        }"
        >{{ props.name }}</span
      >
      <AppButton :class="['button', 'item__button--delete']" label="X" />
    </div>
  </div>
</template>

<style scoped lang="scss">
.todo-item {
  display: flex;
  gap: 10px;
  align-items: center;
  padding: 15px 10px;
  border-bottom: 1px solid #eee;

  .button {
    visibility: hidden;
  }

  &:hover {
    background-color: #f1f1f1;

    .item__button--delete {
      visibility: visible;
    }
  }
}

.todo-item-completed {
  text-decoration-line: line-through;
  color: #808080;
}
</style>
