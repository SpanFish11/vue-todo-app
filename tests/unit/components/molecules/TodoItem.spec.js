import { beforeEach, describe, expect, it } from '@jest/globals'
import { mount, shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import TodoItem from '@/components/molecules/TodoItem.vue'

describe('Given the TodoItem component', () => {
  const completedTodo = { title: 'title', completed: true, id: 'id' }
  const uncompletedTodo = { title: 'title', completed: false, id: 'id' }

  let wrapper
  let store

  describe('when it is rendered', () => {
    beforeEach(() => {
      store = createStore({})
      store.dispatch = jest.fn()
    })

    it.each([[completedTodo], [uncompletedTodo]])('should match the snapshot', (todo) => {
      wrapper = shallowMount(TodoItem, {
        global: {
          plugins: [store],
        },
        props: {
          todo: todo,
        },
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('should dispatch', () => {
    beforeEach(() => {
      store = createStore({})
      store.dispatch = jest.fn()

      wrapper = mount(TodoItem, {
        global: {
          plugins: [store],
        },
        props: {
          todo: uncompletedTodo,
        },
      })
    })

    it('toggleTodo when click on title', async () => {
      await wrapper.find('span').trigger('click')
      await wrapper.vm.$nextTick()

      expect(store.dispatch.mock.calls).toHaveLength(1)
      expect(store.dispatch).toHaveBeenCalledWith('toggleTodo', uncompletedTodo)
    })

    it('toggleTodo when click on checkbox', async () => {
      await wrapper.find('input').trigger('change')
      await wrapper.vm.$nextTick()

      expect(store.dispatch.mock.calls).toHaveLength(1)
      expect(store.dispatch).toHaveBeenCalledWith('toggleTodo', uncompletedTodo)
    })

    it('removeTodo when click on delete button', async () => {
      await wrapper.find('button').trigger('click')
      await wrapper.vm.$nextTick()

      expect(store.dispatch.mock.calls).toHaveLength(1)
      expect(store.dispatch).toHaveBeenCalledWith('removeTodo', uncompletedTodo)
    })
  })
})
