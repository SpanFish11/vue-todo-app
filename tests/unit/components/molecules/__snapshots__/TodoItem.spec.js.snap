// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TodoItem component when it is rendered should match the snapshot 1`] = `
<div
  class="list-group-item d-flex align-items-center"
>
  <app-checkbox-input-stub
    checked="true"
    class="ms-2"
    disabled="false"
  />
  <span
    class="todo-item-completed ms-4 custom-span"
  >
    title
  </span>
  <app-button-stub
    customclass="button,item__button--delete,btn-close,ms-auto"
    label=""
  />
</div>
`;

exports[`Given the TodoItem component when it is rendered should match the snapshot 2`] = `
<div
  class="list-group-item d-flex align-items-center"
>
  <app-checkbox-input-stub
    checked="false"
    class="ms-2"
    disabled="false"
  />
  <span
    class="ms-4 custom-span"
  >
    title
  </span>
  <app-button-stub
    customclass="button,item__button--delete,btn-close,ms-auto"
    label=""
  />
</div>
`;
